skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldExecutorService.kt

```diff
@@ -0,0 +1,122 @@
+     */
+    enum class SkaffoldExecutionMode(val modeFlag: String) {
+        SINGLE_RUN("run"),
+        DEV("dev")
+    }
+
+    /**
+     * Set of settings to control Skaffold execution, including flags and execution mode.
+     *
+     * @property executionMode Mandatory execution mode for Skaffold, see [SkaffoldExecutionMode].
+     * @property skaffoldConfigurationFilePath Optional, location of the Skaffold YAML
+     *           configuration file. If not provided, default `skaffold.yaml` used.
+     * @property workingDirectory Optional, working directory where Skaffold needs to be launched.
+     *           This is usually set to project working directory.
+     */
+    data class SkaffoldExecutorSettings(
+        val executionMode: SkaffoldExecutionMode,
+        val skaffoldConfigurationFilePath: String? = null,
+        var workingDirectory: File? = null
+    )
+
+    /**
+     * Data object with launched Skaffold process and its command line.
+     *
+     * @property process System process for Skaffold.
+     * @property commandLine Command line used to launch the process.
+     */
+    data class SkaffoldProcess(val process: Process, val commandLine: String)
+}
+
+/**
+ * Abstract implementation of [SkaffoldExecutorService], constructs Skaffold command line and
+ * build resulting process. Path to Skaffold executable must be set by subclasses.
+ */
+abstract class AbstractSkaffoldExecutorService : SkaffoldExecutorService {
+    /** Path for Skaffold executable, any form supported by [ProcessBuilder] */
+    protected abstract var skaffoldExecutablePath: Path
+
+    override fun executeSkaffold(settings: SkaffoldExecutorService.SkaffoldExecutorSettings):
+        SkaffoldExecutorService.SkaffoldProcess {
+        val commandList = mutableListOf<String>()
+        with(commandList) {
+            add(skaffoldExecutablePath.toString())
+            add(settings.executionMode.modeFlag)
+            settings.skaffoldConfigurationFilePath?.let {
+                add("--filename")
+                add(it)
+            }
+        }
+
+        return SkaffoldExecutorService.SkaffoldProcess(
+            createProcess(settings.workingDirectory, commandList),
+            commandLine = commandList.joinToString(" ")
+        )
+    }
+
+    @VisibleForTesting
+    fun createProcess(
+        workingDirectory: File?,
+        commandList: List<String>
+    ): Process {
+        val processBuilder = ProcessBuilder()
+        workingDirectory?.let { processBuilder.directory(it) }
+
+        return processBuilder.command(commandList).start()
+    }
+}
+
+/**
+ * Default implementation of Skaffold execution service, where Skaffold executable is assumed to
+ * be already installed on the system and be available in PATH.
+ */
+class DefaultSkaffoldExecutorService : AbstractSkaffoldExecutorService() {
+    // use executable available in PATH
+    override var skaffoldExecutablePath: Path = Paths.get("skaffold")
```

Comment:
Good option. Hard to foresee anything except default / managed Skaffold options as of now, but if we add something else (like pure `kubectl` or some other deployment helper tools), this may get messy and subclassing might be handy. I agree, let's leave this  as it is and keep this option in mind.